### PR TITLE
PR: Highlight commit message

### PR DESCRIPTION
**Pull request by azarkevich**
I recreate pull request, because previous (#2994) was stalled for 1 year after fix some issues...

1. Add ability for specify syntax highlighting for GitExtensions editor via command line:
   `GitExtensions.exe fileeditor --syntax Python <path>`
2. Add simple coloring scheme `GitCommit` which has only definition for comment after #
3. Make default GitExtensions editor command line with syntax 'GitCommit'

Rebase example:

image

### CODE REVIEW

**Comment by orgads** [comment]
The first lines (until a blank line) should be bold, or have some color. This is the commit's subject. It is typically a single line, but is not limited to that. It might be worth it to mark footers (lines at the bottom, starting with a word (possibly multiple words joined with dashes) followed by a colon, like `Signed-off-by: <user>` or `Bug: issue 1234`).

See [Qt Creator's highlighter](http://code.qt.io/cgit/qt-creator/qt-creator.git/tree/src/plugins/git/githighlighters.cpp#n38) for reference.

**Comment by ghanique** [comment]
Other GUI's have a single line text box for the Title of the Commit Message
and a multiline text box for the Body of the Commit Message.

**Comment by azarkevich** [comment]
Not sure if this (select first line ...) will be much valuable. Usually commit message edited in other editor (commit dialog) and viewed in other form (Browser).

I made this highlighting mostly for case when interactive rebase shown 2 or more messages for squash. You can see this case on attached screenshot.

So, I not plan to improve highlighting. Here is no big value for me and no value for other (because PR is not merged)

**Comment by orgads** [comment]
> Not sure if this (select first line ...) will be much valuable. Usually commit message edited in other editor (commit dialog) and viewed in other form (Browser).

I'm not sure what you mean.

1. Commit message is also viewed in Git Extensions itself (where only the subject is shown, unless you click a commit), and in git log (where the subject is not highlighted, but you can filter it by --oneline).
2. Even in the browser, the subject is typically bold. See https://github.com/gitextensions/gitextensions/commit/d87d91aa07c4e5db290f89219c4a03c0240163ce for example.

**Comment by azarkevich** [comment]
I mean, that this pull request is only about editor. This editor shown in limited cases, so improve highlighting in **this** editor has no benefits.

File: GitUI/Editor/GitCommitHighlightingStrategy.cs [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+        static bool IsComment(IDocument document, LineSegment line)
+        {
+            for (int i = 0; i < line.Length; i++)
+            {
+                var c = document.GetCharAt(line.Offset + i);
+                if (Char.IsWhiteSpace(c))
+                    continue;
+
+                if (c == '#')
```
Comment (crypto-rsa) [inline GitUI/Editor/GitCommitHighlightingStrategy.cs]: This won't work if the comment character has been changed in git config file (see my commit 58b7c1aef55ad9620d022b054b196d1208fcefb3 where I fixed it for the interactive rebase patch file).

Comment (azarkevich) [inline GitUI/Editor/GitCommitHighlightingStrategy.cs]: Thanks.

I spent 5 minutes and do not find easy way how to obtain GitModule in GitCommitHighlightingStrategy. So, may be I return to this issue later...

File: GitUI/Editor/CustomSyntaxStrategyManager.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+            catch
+            {
+                StrategyTypes = new Dictionary<string, Type>();
+            }
+        }
+
+        public static IHighlightingStrategy TryFindCustomSyntaxStrategy(string syntax)
+        {
+            Type t;
+            if (!StrategyTypes.TryGetValue(syntax.ToUpperInvariant(), out t))
```
Comment (RussKie) [inline GitUI/Editor/CustomSyntaxStrategyManager.cs]: you may get NRE in the current implementation if `syntax` is `null`
if you make the dictionary case-insensitive you will likely get ANE (https://msdn.microsoft.com/en-us/library/bb347013.aspx#Anchor_1).

File: GitUI/Editor/CustomSyntaxStrategyManager.cs [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+                    .ToDictionary(t => t.Name.EndsWith(StrategySuffix) ? t.Name.Substring(0, t.Name.Length - StrategySuffix.Length).ToUpperInvariant() : t.Name.ToUpperInvariant(), t => t)
+                ;
+            }
+            catch
+            {
+                StrategyTypes = new Dictionary<string, Type>();
+            }
+        }
+
+        public static IHighlightingStrategy TryFindCustomSyntaxStrategy(string syntax)
```
Comment (RussKie) [inline GitUI/Editor/CustomSyntaxStrategyManager.cs]: personally I think this method name is misleading. 
generally `TryXxx` methods are additions to `Xxx` methods and their intent not to throw during the operations which will result in failures in `Xxx` methods.
they also have signatures similar to the following: `TryXxx(T someValue, out TRequired expectedValue): bool`

this methods doesn't fit in the above description.
perhaps a more correct name could be `GetSyntaxStrategy()` or `CreateSyntaxStrategy()` or even simpler `Create()`

File: GitUI/Editor/CustomSyntaxStrategyManager.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using ICSharpCode.TextEditor.Document;
+
+namespace GitUI.Editor
+{
+    // simple manager which return highighting strategy implemented in current assembly
+    class CustomSyntaxStrategyManager
```
Comment (RussKie) [inline GitUI/Editor/CustomSyntaxStrategyManager.cs]: looking at the code this doesn't feel like a manager class but like a factory
so personally I would call this `SyntaxStrategyFactory`

File: GitUI/Editor/CustomSyntaxStrategyManager.cs [hunk 5]
```diff
@@ -0,0 +1,40 @@
+﻿using System;
```
Comment (RussKie) [inline GitUI/Editor/CustomSyntaxStrategyManager.cs]: where are unit tests for the functionality?

File: GitUI/Editor/FileViewerInternal.cs [hunk 6]
```diff
@@ -165,7 +165,11 @@ public void SetText(string text, bool isDiff = false)
 
         public void SetHighlighting(string syntax)
         {
```
Comment (RussKie) [inline GitUI/Editor/FileViewerInternal.cs]: arguments must be validated on all public API
if `syntax` is `null`, `""` or `"   "` there is no reason to proceed further
the only thing you may wish to do in this case (before exiting) is to reset any existing highlighting to defaults

File: GitUI/Editor/FileViewer.cs [hunk 7]
```diff
@@ -420,6 +420,12 @@ public void ViewPatch(Func<string> loadPatchText)
             _async.Load(loadPatchText, ViewPatch);
         }
 
+        public void SetHighlighting(string highlightingSyntax)
+        {
+            _highlightingSyntax = highlightingSyntax;
```
Comment (RussKie) [inline GitUI/Editor/FileViewer.cs]: arguments must be validated on all public API

File: GitUI/Editor/FileViewer.cs [hunk 8]
```diff
@@ -615,7 +621,9 @@ private void ResetForText(string fileName)
         {
             Reset(false, true);
 
-            if (fileName == null)
+            if (_highlightingSyntax != null)
+                _internalFileViewer.SetHighlighting(_highlightingSyntax);
```
Comment (RussKie) [inline GitUI/Editor/FileViewer.cs]: are `""` or `"   "` valid?

File: GitUI/Editor/GitCommitHighlightingStrategy.cs [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+                return false;
+            }
+            return false;
+        }
+
+        void AddWord(IDocument document, LineSegment line, int startOffset, int endOffset, bool isComment)
+        {
+            if (startOffset == endOffset)
+                return;
+            var color = new HighlightColor(isComment ? Color.DarkGreen : Color.Black, false, false);
```
Comment (RussKie) [inline GitUI/Editor/GitCommitHighlightingStrategy.cs]: the colors must be configurable, I like my comments silver grey :)

File: GitUI/Editor/GitCommitHighlightingStrategy.cs [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+                if (Char.IsWhiteSpace(c))
+                    continue;
+
+                if (c == '#')
+                    return true;
+
+                return false;
+            }
+            return false;
+        }
```
Comment (RussKie) [inline GitUI/Editor/GitCommitHighlightingStrategy.cs]: unless I am missing something, the method seems awfully complex for the intent
why not do something like:

```
        static bool IsComment(IDocument document, LineSegment line)
        {
			var text = (/* get line text */).Trim();
			if (string.IsNullOrEmpty(text)) 
			{
				return false;
			}
			return text[0] == /* configured comment char */;
        }
```

File: GitUI/Editor/GitCommitHighlightingStrategy.cs [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+        {
+            for (int i = 0; i < lines.Count; i++)
+            {
+                LineSegment line = lines[i];
+                line.Words = new List<TextWord>();
+
+                ColorLineRedGreen(document, line);
+
+                // and tell the text editor to redraw the changed line
+                document.RequestUpdate(new TextAreaUpdate(TextAreaUpdateType.SingleLine, line.LineNumber));
```
Comment (RussKie) [inline GitUI/Editor/GitCommitHighlightingStrategy.cs]: how performant is this calling for each line? can it be called in batches?

File: GitUI/GitUICommands.cs [hunk 12]
```diff
@@ -2107,17 +2109,38 @@ private void RunMergeToolOrConflictCommand(Dictionary<string, string> arguments)
...
+                {
+                    arguments.Add(argsList[i].TrimStart('-'), null);
+                    argsList.RemoveAt(i);
+                }
+
+                // decrement i, so we examine Ith element again because of shift
+                i--;
             }
+
+            // return updated arguemnts list
```
Comment (RussKie) [inline GitUI/GitUICommands.cs]: typo

File: GitUI/GitUICommands.cs [hunk 13]
```diff
@@ -2107,17 +2109,38 @@ private void RunMergeToolOrConflictCommand(Dictionary<string, string> arguments)
                 StartResolveConflictsDialog();
         }
 
-        private static Dictionary<string, string> InitializeArguments(string[] args)
+        private static Dictionary<string, string> InitializeArguments(ref string[] args)
```
Comment (RussKie) [inline GitUI/GitUICommands.cs]: this looks like a code smell. why do you need to modify the `args` array?

**Comment by RussKie** [comment]
Superseded by #4809

**Pull request closed without merging**